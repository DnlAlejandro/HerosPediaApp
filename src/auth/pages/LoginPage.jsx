import { useContext } from "react"
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context/AuthContext"

export const LoginPage = () => {

  const { login } = useContext( AuthContext )
  const navigate = useNavigate()

  const handleLogin = () => {

    const lastpath = localStorage.getItem('lastpath') || '/heroes-app/';

    login('DnlAlejandro');

    navigate(lastpath, {
      replace: true
  })
  }

  return (
    <div className="flex flex-col items-center justify-center h-screen">
      <h1 className="text-3xl font-bold text-gray-900">Login</h1>
      <button
        className="w-full max-w-sm mt-8 bg-indigo-500 hover:bg-indigo-700 rounded-lg shadow-xl font-medium text-white px-4 py-2"
        onClick={handleLogin}
      >
        Join
      </button>
    </div>
  )
}
