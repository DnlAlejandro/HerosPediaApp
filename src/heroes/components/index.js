export { HeroList } from './HeroList';
export { HeroCard } from './HeroCard';