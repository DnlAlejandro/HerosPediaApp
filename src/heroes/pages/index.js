export { DcPage } from './DcPage';
export { MarvelPage } from './MarvelPage';
export { Hero } from './Hero';
export { SearchPage } from './SearchPage';