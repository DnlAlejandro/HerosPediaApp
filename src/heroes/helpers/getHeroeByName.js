import { heroes } from "../data/heroes";

export const getHeroeByName = (name = "") => {
    name = name.toLowerCase().trim();

    if (name.length === 0) return [];

    return heroes.filter((hero) => hero.superhero.toLowerCase().includes(name));
};
