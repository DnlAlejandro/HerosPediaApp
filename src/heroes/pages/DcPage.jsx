import { HeroList } from "../components"

export const DcPage = () => {
  return (
    <div>
      <span className="text-blue-500">Characters</span>
      <h1 className="text-4xl font-bold mb-3 text-blue-700">DC Comics</h1>

      <HeroList
        publisher={'DC Comics'}
      />

    </div>
  )
}
