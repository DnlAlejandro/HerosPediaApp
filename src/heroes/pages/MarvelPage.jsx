import { HeroList } from "../components";

export const MarvelPage = () => {
    return (
        <div>
            <span className="text-red-500">Characters</span>
            <h1 className="text-4xl font-bold mb-3 text-red-700">
                Marvel Comics
            </h1>

            <HeroList publisher={"Marvel Comics"} />
        </div>
    );
};
