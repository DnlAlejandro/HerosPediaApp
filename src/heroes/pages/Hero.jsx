import { useParams } from "react-router-dom";
import { getHeroeById } from "../helpers";
import { Navigate } from "react-router-dom";
import { useNavigate } from "react-router-dom";
import { useMemo } from "react";

export const Hero = () => {
    const { id } = useParams();

    const { superhero, publisher, alter_ego, first_appearance, characters } =
        useMemo(() => getHeroeById(id), [id]);

    const heroImageURL = `/heroes/${id}.jpg`;

    if (!superhero) {
        return <Navigate to="/marvel" />;
    }

    const navigate = useNavigate();

    const handleNavigateBar = () => {
        navigate(-1, {
            replace: true,
        });
    };

    return (
        <>
            <div className="flex gap-5">
                <div>
                    <img
                        src={heroImageURL}
                        className="rounded-lg h-full max-h-[70vh] min-w-72 hover:scale-105"
                        alt={superhero}
                    />
                    <h1 className="mt-4 font-bold text-3xl text-slate-600">
                        {superhero}
                    </h1>
                    <span
                        className={
                            publisher === "Marvel Comics"
                                ? "text-red-700"
                                : "text-blue-600"
                        }
                    >
                        {publisher}
                    </span>
                </div>
                <div className="w-full">
                    {alter_ego === characters ? (
                        <div className="my-5 shadow-md p-5 rounded-md">
                            <h4 className="font-semibold text-2xl text-indigo-600">
                                Alter Ego & Character
                            </h4>
                            <span className="text-sm text-slate-500">
                                {alter_ego}
                            </span>
                        </div>
                    ) : (
                        <div className="*:my-5 *:shadow-md *:p-5 *:rounded-md">
                            <div>
                                <h4 className="font-semibold text-2xl text-indigo-600">
                                    Alter Ego
                                </h4>
                                <span className="text-sm text-slate-500">
                                    {alter_ego}
                                </span>
                            </div>
                            <div>
                                <h4 className="font-semibold text-2xl text-indigo-600">
                                    Characters
                                </h4>
                                <span className="text-sm text-slate-500">
                                    {characters}
                                </span>
                            </div>
                        </div>
                    )}
                    <div className="shadow-md p-5 rounded-md">
                        <h4
                            className={`font-semibold text-2xl ${
                                publisher === "Marvel Comics"
                                    ? "text-red-700"
                                    : "text-blue-600"
                            }`}
                        >
                            First Appearance
                        </h4>
                        <span className="text-sm text-slate-500">
                            {first_appearance}
                        </span>
                    </div>
                    <button
                        className="bg-indigo-600 mt-4 text-white text-sm w-full h-12 rounded-md hover:bg-indigo-700"
                        onClick={handleNavigateBar}
                        data-testid="goback-btn"
                    >
                        Go back
                    </button>
                </div>
            </div>
        </>
    );
};
