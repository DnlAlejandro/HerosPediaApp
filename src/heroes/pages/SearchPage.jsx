import { useLocation } from "react-router-dom";
import { HeroCard } from "../components";
import queryString from "query-string";
import { getHeroeByName } from "../helpers";

export const SearchPage = () => {
    const location = useLocation();

    const { q = "" } = queryString.parse(location.search);

    const heroes = getHeroeByName(q);

    const searchAlert = (
        <div
            data-testid="search-info"
            className="mt-5 bg-blue-100 p-5 rounded-lg text-blue-600 border border-blue-300/40 flex items-center gap-3 "
        >
            <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={1.5}
                stroke="currentColor"
                className="w-6 h-6"
            >
                <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    d="m21 21-5.197-5.197m0 0A7.5 7.5 0 1 0 5.196 5.196a7.5 7.5 0 0 0 10.607 10.607Z"
                />
            </svg>
            <p className="font-light">Search a hero</p>
        </div>
    );

    const errorAlert = (
        <div
            data-testid="notfound-info"
            className="mt-5 bg-yellow-100 p-5 rounded-lg text-yellow-600 border border-yellow-300/40 flex items-center gap-3"
        >
            <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={1.5}
                stroke="currentColor"
                className="w-6 h-6"
            >
                <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    d="M12 9v3.75m9-.75a9 9 0 1 1-18 0 9 9 0 0 1 18 0Zm-9 3.75h.008v.008H12v-.008Z"
                />
            </svg>
            <p className="font-light">
                There are not results with{" "}
                <span className="font-medium">{q}</span>
            </p>
        </div>
    );

    return (
        <>
            <h1 className="text-3xl font-bold text-indigo-600">Results</h1>
            {heroes.length !== 0 && (
                <span className="text-md text-slate-600">{q}</span>
            )}

            {q === "" ? searchAlert : heroes.length === 0 && errorAlert}

            <div className="flex gap-4 flex-wrap">
                {heroes.map((hero) => (
                    <HeroCard
                        key={hero.id}
                        id={hero.id}
                        superhero={hero.superhero}
                    />
                ))}
            </div>
        </>
    );
};
