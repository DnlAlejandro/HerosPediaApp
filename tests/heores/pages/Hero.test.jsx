import { render, screen, fireEvent } from "@testing-library/react";
import { Hero } from "../../../src/heroes/pages/Hero";
import { MemoryRouter } from "react-router-dom";
import { getHeroeById } from "../../../src/heroes/helpers";

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
}));

jest.mock('../../../src/heroes/helpers/getHeroeById')

describe('Test in componente <Hero/>', () => {

    beforeEach(() => jest.resetAllMocks());

    test('should match with snapshot', () => {

        getHeroeById.mockReturnValue({
            'id': 'dc-batman',
            'superhero': 'Batman',
            'publisher': 'DC Comics',
            'alter_ego': 'Bruce Wayne',
            'first_appearance': 'Detective Comics #27',
            'characters': 'Bruce Wayne'
        })

        const { container } = render(
            <MemoryRouter initialEntries={[`/hero/dc-batman`]}>
                <Hero />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();

    });

    test('when button is clicked should return once', () => {

        getHeroeById.mockReturnValue({
            'id': 'dc-batman',
            'superhero': 'Batman',
            'publisher': 'DC Comics',
            'alter_ego': 'Bruce Wayne',
            'first_appearance': 'Detective Comics #27',
            'characters': 'Bruce Wayne'
        })

        const { container } = render(
            <MemoryRouter initialEntries={[`/hero/dc-batman`]}>
                <Hero />
            </MemoryRouter>
        )

        const goBack = screen.getByTestId("goback-btn");
        fireEvent.click(goBack);
        
        expect( mockUseNavigate ).toHaveBeenCalledWith(-1, { replace: true })
    });
});