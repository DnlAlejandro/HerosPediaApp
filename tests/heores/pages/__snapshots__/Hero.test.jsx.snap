// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in componente <Hero/> debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="flex gap-5"
  >
    <div>
      <img
        alt="Batman"
        class="rounded-lg h-full max-h-[70vh] min-w-72 hover:scale-105"
        src="/heroes/undefined.jpg"
      />
      <h1
        class="mt-4 font-bold text-3xl text-slate-600"
      >
        Batman
      </h1>
      <span
        class="text-blue-600"
      >
        DC Comics
      </span>
    </div>
    <div
      class="w-full"
    >
      <div
        class="my-5 shadow-md p-5 rounded-md"
      >
        <h4
          class="font-semibold text-2xl text-indigo-600"
        >
          Alter Ego & Character
        </h4>
        <span
          class="text-sm text-slate-500"
        >
          Bruce Wayne
        </span>
      </div>
      <div
        class="shadow-md p-5 rounded-md"
      >
        <h4
          class="font-semibold text-2xl text-blue-600"
        >
          First Appearance
        </h4>
        <span
          class="text-sm text-slate-500"
        >
          Detective Comics #27
        </span>
      </div>
      <button
        class="bg-indigo-600 mt-4 text-white text-sm w-full h-12 rounded-md hover:bg-indigo-700"
        data-testid="goback-btn"
      >
        Go back
      </button>
    </div>
  </div>
</div>
`;

exports[`Test in componente <Hero/> should match con el snapshot 1`] = `
<div>
  <div
    class="flex gap-5"
  >
    <div>
      <img
        alt="Batman"
        class="rounded-lg h-full max-h-[70vh] min-w-72 hover:scale-105"
        src="/heroes/undefined.jpg"
      />
      <h1
        class="mt-4 font-bold text-3xl text-slate-600"
      >
        Batman
      </h1>
      <span
        class="text-blue-600"
      >
        DC Comics
      </span>
    </div>
    <div
      class="w-full"
    >
      <div
        class="my-5 shadow-md p-5 rounded-md"
      >
        <h4
          class="font-semibold text-2xl text-indigo-600"
        >
          Alter Ego & Character
        </h4>
        <span
          class="text-sm text-slate-500"
        >
          Bruce Wayne
        </span>
      </div>
      <div
        class="shadow-md p-5 rounded-md"
      >
        <h4
          class="font-semibold text-2xl text-blue-600"
        >
          First Appearance
        </h4>
        <span
          class="text-sm text-slate-500"
        >
          Detective Comics #27
        </span>
      </div>
      <button
        class="bg-indigo-600 mt-4 text-white text-sm w-full h-12 rounded-md hover:bg-indigo-700"
        data-testid="goback-btn"
      >
        Go back
      </button>
    </div>
  </div>
</div>
`;

exports[`Test in componente <Hero/> should match wit snapshot 1`] = `
<div>
  <div
    class="flex gap-5"
  >
    <div>
      <img
        alt="Batman"
        class="rounded-lg h-full max-h-[70vh] min-w-72 hover:scale-105"
        src="/heroes/undefined.jpg"
      />
      <h1
        class="mt-4 font-bold text-3xl text-slate-600"
      >
        Batman
      </h1>
      <span
        class="text-blue-600"
      >
        DC Comics
      </span>
    </div>
    <div
      class="w-full"
    >
      <div
        class="my-5 shadow-md p-5 rounded-md"
      >
        <h4
          class="font-semibold text-2xl text-indigo-600"
        >
          Alter Ego & Character
        </h4>
        <span
          class="text-sm text-slate-500"
        >
          Bruce Wayne
        </span>
      </div>
      <div
        class="shadow-md p-5 rounded-md"
      >
        <h4
          class="font-semibold text-2xl text-blue-600"
        >
          First Appearance
        </h4>
        <span
          class="text-sm text-slate-500"
        >
          Detective Comics #27
        </span>
      </div>
      <button
        class="bg-indigo-600 mt-4 text-white text-sm w-full h-12 rounded-md hover:bg-indigo-700"
        data-testid="goback-btn"
      >
        Go back
      </button>
    </div>
  </div>
</div>
`;

exports[`Test in componente <Hero/> should match with snapshot 1`] = `
<div>
  <div
    class="flex gap-5"
  >
    <div>
      <img
        alt="Batman"
        class="rounded-lg h-full max-h-[70vh] min-w-72 hover:scale-105"
        src="/heroes/undefined.jpg"
      />
      <h1
        class="mt-4 font-bold text-3xl text-slate-600"
      >
        Batman
      </h1>
      <span
        class="text-blue-600"
      >
        DC Comics
      </span>
    </div>
    <div
      class="w-full"
    >
      <div
        class="my-5 shadow-md p-5 rounded-md"
      >
        <h4
          class="font-semibold text-2xl text-indigo-600"
        >
          Alter Ego & Character
        </h4>
        <span
          class="text-sm text-slate-500"
        >
          Bruce Wayne
        </span>
      </div>
      <div
        class="shadow-md p-5 rounded-md"
      >
        <h4
          class="font-semibold text-2xl text-blue-600"
        >
          First Appearance
        </h4>
        <span
          class="text-sm text-slate-500"
        >
          Detective Comics #27
        </span>
      </div>
      <button
        class="bg-indigo-600 mt-4 text-white text-sm w-full h-12 rounded-md hover:bg-indigo-700"
        data-testid="goback-btn"
      >
        Go back
      </button>
    </div>
  </div>
</div>
`;
